### PR TITLE
Account for exceptional closures in requesting date logic

### PR DESCRIPTION
**Pull request by davidpmccormick**
Up till now, we haven't taken any e.g. bank holidays into account when determining the next available pick-up date for requested items.

With this change, if a request is made after 10am on a Friday and the following Monday is a bank holiday, the next available pick-up date will be the following Wednesday (at present, it would be the Tuesday).

### CODE REVIEW

File: catalogue/webapp/test/utils/dates.test.ts [hunk 1]
```diff
@@ -182,26 +196,36 @@ describe('determineNextAvailableDate', () => {
...
     expect(result4).toEqual(new Date('2022-09-08T10:30:00+0100'));
   });
+
+  it('accounts for exceptional closure dates', () => {
+    const exceptionalClosure = new Date('2021-12-13'); // Monday
+    const result = determineNextAvailableDate(
+      new Date('2021-12-10 10:30'), // Friday
+      [0],
+      [exceptionalClosure]
+    ); // Sunday
```
Comment (rcantin-w) [inline catalogue/webapp/test/utils/dates.test.ts:226]: Why Sunday here? Everything looks logical, just wanting to understand the usual logic for these!

Comment (davidpmccormick) [reply catalogue/webapp/test/utils/dates.test.ts:226]: You're right, that `// Sunday` shouldn't be there. The result of a copy-paste. I'll get rid of it, thanks.

File: catalogue/webapp/test/utils/dates.test.ts [hunk 2]
```diff
@@ -182,26 +196,36 @@ describe('determineNextAvailableDate', () => {
     // at 09:30 in London -- it can be fulfilled the next day.
     const date1 = new Date('2021-12-09T10:30:00+0100');
 
-    const result1 = determineNextAvailableDate(date1, [0]);
+    const result1 = determineNextAvailableDate(date1, [0], []);
     expect(result1).toEqual(new Date('2021-12-10T09:30:00Z'));
 
     // Paris is an hour ahead of London, so a request made at 11:30 in Paris is
     // at 19:30 in London -- it can’t be fulfilled the next day.
```
Comment (rcantin-w) [inline catalogue/webapp/test/utils/dates.test.ts:203]: Should this be 10:30 in London?

Comment (alexwlchan) [reply catalogue/webapp/test/utils/dates.test.ts:203]: Yes, that's my mistake.

Comment (davidpmccormick) [reply catalogue/webapp/test/utils/dates.test.ts:203]: definitely!

**Review by rcantin-w — APPROVED** [review]
Apart from the maybe typo in the comment I think it looks good, and thoroughly tested 🤞 👍

File: catalogue/webapp/utils/dates.ts [hunk 3]
```diff
@@ -73,33 +74,50 @@ export function findNextPickUpDay(
...
   const nextDay = addDays(date, 1);
-  const isClosedThisDay = regularClosedDays.includes(
-    date.getDay() as DayNumber
-  );
-  const isOpenNextDay = !regularClosedDays.includes(
-    nextDay.getDay() as DayNumber
-  );
+  const isClosedThisDay =
+    regularClosedDays.includes(date.getDay() as DayNumber) ||
+    exceptionalClosedDates.find(d => d.toDateString() === date.toDateString());
```
Comment (davidpmccormick) [inline catalogue/webapp/utils/dates.ts:79]: This should use `isSameDay` with a `mode` of 'London' once #8455 is in

File: catalogue/webapp/utils/dates.ts [hunk 4]
```diff
@@ -73,33 +74,50 @@ export function findNextPickUpDay(
...
+  const isOpenNextDay =
+    !regularClosedDays.includes(nextDay.getDay() as DayNumber) &&
+    !exceptionalClosedDates.find(
+      d => d.toDateString() === nextDay.toDateString()
```
Comment (davidpmccormick) [inline catalogue/webapp/utils/dates.ts:83]: …and this

**Review by alexwlchan — APPROVED** [review]
Looks good, modulo `isSameDay()` changes.

**Comment by alexwlchan** [comment]
Merging so we can get this deployed today.

**Pull request merged**